smt.go

```diff
@@ -341,27 +400,105 @@ func (tree *BASSparseMerkleTree) Set(key uint64, val []byte) error {
 	for i := 0; i < int(tree.maxDepth)/4; i++ {
 		path := key >> (int(tree.maxDepth) - (i+1)*4)
 		nibble := path & 0x000000000000000f
-		parentNodes = append(parentNodes, targetNode)
+		parentNodes = append(parentNodes, targetNode.Copy())
 		if err := tree.extendNode(targetNode, nibble, path, depth, true); err != nil {
 			return err
 		}
 		targetNode = targetNode.Children[nibble]
 
 		depth += 4
 	}
-	targetNode = targetNode.set(val, newVersion) // leaf
-	tree.journal[journalKey{targetNode.depth, targetNode.path}] = targetNode
-	// recompute root hash
+	targetNode = targetNode.Copy()
+	targetNode.Set(val, newVersion) // update hash of leaf node
+	tree.journal.set(journalKey{targetNode.depth, targetNode.path}, targetNode)
+	// recompute root hash of middle nodes
 	for i := len(parentNodes) - 1; i >= 0; i-- {
 		childNibble := key >> (int(tree.maxDepth) - (i+1)*4) & 0x000000000000000f
-		targetNode = parentNodes[i].setChildren(targetNode, int(childNibble), newVersion)
+		parentNodes[i].SetChildren(targetNode, int(childNibble), newVersion)
 
-		tree.journal[journalKey{targetNode.depth, targetNode.path}] = targetNode
+		targetNode = parentNodes[i]
+		tree.journal.set(journalKey{targetNode.depth, targetNode.path}, targetNode)
 	}
 	tree.root = targetNode
 	return nil
 }
 
+func (tree *BASSparseMerkleTree) MultiSet(items ...Item) error {
```

Comment:
Why the parameter is `...Item`, does it make sense if nothing to set at all? If not, let us change to slice.

Reply:
updated